File: src/main/java/com/portofino/realtrainmodunofficial/rail/math/BezierCurve.java
package com.portofino.realtrainmodunofficial.rail.math;

import net.minecraft.util.Mth;

/**
 * Port of jp.legacy.legacylib.math.BezierCurve (cubic Bezier on Z/X plane).
 */
public final class BezierCurve implements ILine {
    public static final int QUANTIZE = 32;
    /** レンダラの {@code max} と同一にし、{@code max > split} による重複サンプル（スパイク）を防ぐ。 */
    public static final int MAX_CURVE_RENDER_SPLIT = 384;

    public final double[] sp;
    public final double[] cpS;
    public final double[] cpE;
    public final double[] ep;
    private float[] normalizedParameters;
    private final double length;
    private final int split;

    /**
     * 弧長に応じた分割数。{@link com.portofino.realtrainmodunofficial.rail.util.RailMap#curveSplitForLength(double)} と同じ値を返す。
     */
    public static int splitForLength(double arcLength) {
        if (arcLength < 1.0e-4) {
            return 32;
        }
        int raw = (int) (arcLength * 32.0);
        int s = Mth.clamp(Math.max(raw, 8), 8, 4096);
        return Math.min(s, MAX_CURVE_RENDER_SPLIT);
    }

    public BezierCurve(double p1, double p2, double p3, double p4, double p5, double p6, double p7, double p8) {
        this.sp = new double[]{p1, p2};
        this.cpS = new double[]{p3, p4};
        this.cpE = new double[]{p5, p6};
        this.ep = new double[]{p7, p8};
        this.length = this.calcLength();
        this.split = splitForLength(this.length);
    }

    @Override
    public double[] getPoint(int par1, int par2) {
        return this.getPointFromParameter((double) this.getHomogenizedParameter(par1, par2));
    }

    private double[] getPointFromParameter(double par1) {
        double t = par1 < 0.0D ? 0.0D : (par1 > 1.0D ? 1.0D : par1);
        double tp = 1.0D - t;
        double d0 = t * t * t;
        double d1 = 3.0D * t * t * tp;
        double d2 = 3.0D * t * tp * tp;
        double d3 = tp * tp * tp;
        double x = d0 * this.ep[0] + d1 * this.cpE[0] + d2 * this.cpS[0] + d3 * this.sp[0];
        double y = d0 * this.ep[1] + d1 * this.cpE[1] + d2 * this.cpS[1] + d3 * this.sp[1];
        return new double[]{x, y};
    }

    @Override
    public int getNearlestPoint(int par1, double par2, double par3) {
        int i = 0;
        double pd = Double.MAX_VALUE;
        for (int j = 0; j < par1; ++j) {
            double[] point = this.getPoint(par1, j);
            double dx = par2 - point[1];
            double dy = par3 - point[0];
            double distance = dx * dx + dy * dy;
            if (distance < pd) {
                pd = distance;
                i = j;
            }
        }
        return pd < Double.MAX_VALUE ? i : -1;
    }

    @Override
    public double getSlope(int par1, int par2) {
        return this.getSlopeFromParameter((double) this.getHomogenizedParameter(par1, par2));
    }

    private double getSlopeFromParameter(double par1) {
        double t = par1 < 0.0D ? 0.0D : (par1 > 1.0D ? 1.0D : par1);
        double tp = 1.0D - t;
        double d0 = t * t;
        double d1 = 2.0D * t * tp;
        double d2 = tp * tp;
        double dx = 3.0D * (d0 * (this.ep[0] - this.cpE[0]) + d1 * (this.cpE[0] - this.cpS[0]) + d2 * (this.cpS[0] - this.sp[0]));
        double dy = 3.0D * (d0 * (this.ep[1] - this.cpE[1]) + d1 * (this.cpE[1] - this.cpS[1]) + d2 * (this.cpS[1] - this.sp[1]));
        return Math.atan2(dy, dx);
    }

    private float getHomogenizedParameter(int n, int par2) {
        if (n < 4) {
            return 0.0F;
        } else if (par2 <= 0) {
            return 0.0F;
        } else if (par2 >= n) {
            return 1.0F;
        } else {
            if (this.normalizedParameters == null) {
                this.initNP();
            }
            int i0 = CurveMath.floor((float) par2 * (float) this.split / (float) n);
            if (i0 < 0) i0 = 0;
            if (i0 >= this.normalizedParameters.length) i0 = this.normalizedParameters.length - 1;
            return this.normalizedParameters[i0];
        }
    }

    private void initNP() {
        if (this.split < 1) {
            this.normalizedParameters = new float[]{0.0F};
            return;
        }
        this.normalizedParameters = new float[this.split];
        float ni = 1.0F / (float) this.split;
        float tt = 0.0F;
        double[] p = this.sp;
        double[] q = new double[2];
        float[] dd = new float[this.split + 1];
        dd[0] = 0.0F;
        int i;
        for (i = 1; i < this.split + 1; ++i) {
            tt += ni;
            q = this.getPointFromParameter((double) tt);
            dd[i] = dd[i - 1] + (float) this.getDistance(p[0], q[0], p[1], q[1]);
            p = q;
        }
        float total = dd[this.split];
        if (total < 1.0e-8f) {
            for (i = 0; i < this.split; ++i) {
                this.normalizedParameters[i] = (float) i / (float) this.split;
            }
            return;
        }
        for (i = 1; i < this.split + 1; ++i) {
            dd[i] /= total;
        }
        for (i = 0; i < this.split; ++i) {
            float t = (float) i / (float) this.split;
            int k;
            for (k = 0; k < this.split - 1 && (!(dd[k] <= t) || !(t <= dd[k + 1])); ++k) {
            }
            float denom = dd[k + 1] - dd[k];
            float x;
            if (Math.abs(denom) < 1.0e-8f) {
                x = (float) k / (float) this.split;
            } else {
                x = (t - dd[k]) / denom;
                x = ((float) k * (1.0F - x) + (float) (1 + k) * x) * (1.0F / (float) this.split);
            }
            this.normalizedParameters[i] = x;
        }
    }

    @Override
    public double getLength() {
        return this.length;
    }

    private double calcLength() {
        double x0 = this.sp[0] - this.ep[0];
        double y0 = this.sp[1] - this.ep[1];
        double l0 = Math.sqrt(x0 * x0 + y0 * y0);
        int n = CurveMath.floor(l0 * 2.0D);
        if (n < 1) n = 1;
        float ni = 1.0F / (float) n;
        float tt = 0.0F;
        double[] p = this.sp;
        double[] q = new double[2];
        double[] dd = new double[n + 1];
        dd[0] = 0.0D;
        for (int i = 1; i < n + 1; ++i) {
            tt += ni;
            q = this.getPointFromParameter((double) tt);
            dd[i] = dd[i - 1] + this.getDistance(p[0], q[0], p[1], q[1]);
            p = q;
        }
        return dd[n];
    }

    private double getDistance(double par1, double par2, double par3, double par4) {
        double xDis = Math.abs(par1 - par2);
        double yDis = Math.abs(par3 - par4);
        return Math.sqrt(xDis * xDis + yDis * yDis);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/rail/math/BezierCurve.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.rail.math

import cc.mirukuneko.realtrainmodrenewed.rail.util.RailMap
import net.minecraft.util.Mth
import kotlin.math.*

/**
 * Port of jp.legacy.legacylib.math.BezierCurve (cubic Bezier on Z/X plane).
 */
class BezierCurve(
    p1: Double, p2: Double, p3: Double, p4: Double,
    p5: Double, p6: Double, p7: Double, p8: Double
) : ILine {
    companion object {
        const val QUANTIZE: Int = 32
        /** レンダラの max と同一にし、max > split による重複サンプル（スパイク）を防ぐ。 */
        const val MAX_CURVE_RENDER_SPLIT: Int = 384

        /**
         * 弧長に応じた分割数。[RailMap.curveSplitForLength] と同じ値を返す。
         */
        @JvmStatic
        fun splitForLength(arcLength: Double): Int {
            if (arcLength < 1.0e-4) return 32
            val raw = (arcLength * 32.0).toInt()
            val s = Mth.clamp(maxOf(raw, 8), 8, 4096)
            return minOf(s, MAX_CURVE_RENDER_SPLIT)
        }
    }

    @JvmField val sp: DoubleArray = doubleArrayOf(p1, p2)
    @JvmField val cpS: DoubleArray = doubleArrayOf(p3, p4)
    @JvmField val cpE: DoubleArray = doubleArrayOf(p5, p6)
    @JvmField val ep: DoubleArray = doubleArrayOf(p7, p8)

    private var normalizedParameters: FloatArray? = null
    override val length: Double
    val split: Int

    init {
        length = calcLength()
        split = splitForLength(length)
    }

    override fun getPoint(par1: Int, par2: Int): DoubleArray =
        getPointFromParameter(getHomogenizedParameter(par1, par2).toDouble())

    private fun getPointFromParameter(par1: Double): DoubleArray {
        val t = par1.coerceIn(0.0, 1.0)
        val tp = 1.0 - t
        val d0 = t * t * t
        val d1 = 3.0 * t * t * tp
        val d2 = 3.0 * t * tp * tp
        val d3 = tp * tp * tp
        val x = d0 * ep[0] + d1 * cpE[0] + d2 * cpS[0] + d3 * sp[0]
        val y = d0 * ep[1] + d1 * cpE[1] + d2 * cpS[1] + d3 * sp[1]
        return doubleArrayOf(x, y)
    }

    override fun getNearlestPoint(par1: Int, par2: Double, par3: Double): Int {
        var i = 0
        var pd = Double.MAX_VALUE
        for (j in 0 until par1) {
            val point = getPoint(par1, j)
            val dx = par2 - point[1]
            val dy = par3 - point[0]
            val distance = dx * dx + dy * dy
            if (distance < pd) {
                pd = distance
                i = j
            }
        }
        return if (pd < Double.MAX_VALUE) i else -1
    }

    override fun getSlope(par1: Int, par2: Int): Double =
        getSlopeFromParameter(getHomogenizedParameter(par1, par2).toDouble())

    private fun getSlopeFromParameter(par1: Double): Double {
        val t = par1.coerceIn(0.0, 1.0)
        val tp = 1.0 - t
        val d0 = t * t
        val d1 = 2.0 * t * tp
        val d2 = tp * tp
        val dx = 3.0 * (d0 * (ep[0] - cpE[0]) + d1 * (cpE[0] - cpS[0]) + d2 * (cpS[0] - sp[0]))
        val dy = 3.0 * (d0 * (ep[1] - cpE[1]) + d1 * (cpE[1] - cpS[1]) + d2 * (cpS[1] - sp[1]))
        return atan2(dy, dx)
    }

    private fun getHomogenizedParameter(n: Int, par2: Int): Float {
        return when {
            n < 4 -> 0.0f
            par2 <= 0 -> 0.0f
            par2 >= n -> 1.0f
            else -> {
                val np = normalizedParameters ?: initNP().also { normalizedParameters = it }
                val i0 = CurveMath.floor(par2.toFloat() * split.toFloat() / n.toFloat())
                    .coerceIn(0, np.size - 1)
                np[i0]
            }
        }
    }

    private fun initNP(): FloatArray {
        if (split < 1) return floatArrayOf(0.0f)
        val np = FloatArray(split)
        val ni = 1.0f / split.toFloat()
        var tt = 0.0f
        var p = sp
        var q: DoubleArray
        val dd = FloatArray(split + 1)
        dd[0] = 0.0f
        for (i in 1..split) {
            tt += ni
            q = getPointFromParameter(tt.toDouble())
            dd[i] = dd[i - 1] + getDistance(p[0], q[0], p[1], q[1]).toFloat()
            p = q
        }
        val total = dd[split]
        if (total < 1.0e-8f) {
            for (i in 0 until split) {
                np[i] = i.toFloat() / split.toFloat()
            }
            return np
        }
        for (i in 1..split) {
            dd[i] /= total
        }
        for (i in 0 until split) {
            val t = i.toFloat() / split.toFloat()
            var k = 0
            while (k < split - 1 && !(dd[k] <= t && t <= dd[k + 1])) k++
            val denom = dd[k + 1] - dd[k]
            val x: Float = if (abs(denom) < 1.0e-8f) {
                k.toFloat() / split.toFloat()
            } else {
                ((t - dd[k]) / denom).let { v ->
                    (k.toFloat() * (1.0f - v) + (1 + k).toFloat() * v) * (1.0f / split.toFloat())
                }
            }
            np[i] = x
        }
        return np
    }

    private fun calcLength(): Double {
        val x0 = sp[0] - ep[0]
        val y0 = sp[1] - ep[1]
        val l0 = sqrt(x0 * x0 + y0 * y0)
        var n = CurveMath.floor(l0 * 2.0)
        if (n < 1) n = 1
        val ni = 1.0f / n.toFloat()
        var tt = 0.0f
        var p = sp
        var q: DoubleArray
        val dd = DoubleArray(n + 1)
        dd[0] = 0.0
        for (i in 1..n) {
            tt += ni
            q = getPointFromParameter(tt.toDouble())
            dd[i] = dd[i - 1] + getDistance(p[0], q[0], p[1], q[1])
            p = q
        }
        return dd[n]
    }

    private fun getDistance(par1: Double, par2: Double, par3: Double, par4: Double): Double {
        val xDis = abs(par1 - par2)
        val yDis = abs(par3 - par4)
        return sqrt(xDis * xDis + yDis * yDis)
    }
}
